File: digest_system/runtime/artifacts.py
"""Shared stage plumbing: paths, run-directory bookkeeping, Markdown extraction.

This is the Python port of ``src/runtime/artifacts.mjs``. It keeps the utilities every
stage needs to read, write and validate its artifacts. Transport, retry policy and cost
accounting live in their own modules.

One deliberate difference from the JavaScript runner: the repository root is resolved
once and passed explicitly rather than by changing the process's working directory. The
JavaScript runner called ``process.chdir(runDirectory)``, which is a hidden global
dependency; here every path is absolute and derived from :data:`ROOT`.
"""

from __future__ import annotations

import json
import os
import re
import shutil
from pathlib import Path
from typing import Any, Iterable, Sequence

#: Repository root. Resolved from this file's location (``digest_system/runtime/``), and
#: overridable with ``DIGEST_ROOT`` so tests can point at a temporary tree.
ROOT = Path(os.environ.get("DIGEST_ROOT") or Path(__file__).resolve().parents[2]).resolve()

RUNS_DIRECTORY = ".digest-runs"

#: Reasoning tokens count against ``max_tokens`` on the DeepSeek model. A measured replay
#: showed edit stages spending 28-31K reasoning tokens on a small corpus, which left only
#: ~1.5K for the artifact and silently truncated it. The cap must therefore cover reasoning
#: plus the full artifact. DeepSeek's maximum is 384K.
MAX_OUTPUT_TOKENS = int(os.environ.get("DIGEST_MAX_OUTPUT_TOKENS", 262_144))


class RunnerError(RuntimeError):
    """A pipeline error that should be reported to the operator, not a crash."""


def runs_root(root: Path | None = None) -> Path:
    return (root or ROOT) / RUNS_DIRECTORY


def run_directory(run_id: str, root: Path | None = None) -> Path:
    return runs_root(root) / run_id


def stage_directory(run_id: str, stage_name: str, root: Path | None = None) -> Path:
    """The canonical stage directory for a run, shared by the orchestrator and readers."""
    return run_directory(run_id, root) / stage_name


def validate_run_id(run_id: str) -> None:
    if not run_id or run_id in {".", ".."} or Path(run_id).name != run_id:
        raise RunnerError("Run ID must be a single directory name")


def exists(file_path: Path) -> bool:
    """Whether a path exists, for a file *or* a directory.

    The JavaScript version used ``readFile``, which fails on a directory, so two guards
    never fired. This asks the filesystem a question that has an answer for both kinds of
    path.
    """
    return file_path.exists()


def read_json(file_path: Path) -> Any:
    return json.loads(file_path.read_text(encoding="utf-8"))


def try_read_json(file_path: Path) -> Any | None:
    try:
        return read_json(file_path)
    except (OSError, ValueError):
        return None


def copy_file(source: Path, destination: Path) -> None:
    destination.parent.mkdir(parents=True, exist_ok=True)
    shutil.copyfile(source, destination)


def write_artifact(file_path: Path, content: str) -> None:
    """Write an artifact, recreating its directory if it has disappeared.

    Used for failure records: a run directory can be deleted while a run is in flight, and
    the failure path must not then replace the error being recorded with an ``ENOENT``
    about the record itself.
    """
    file_path.parent.mkdir(parents=True, exist_ok=True)
    file_path.write_text(content, encoding="utf-8")


def write_json(file_path: Path, value: Any) -> None:
    write_artifact(file_path, json.dumps(value, ensure_ascii=False, indent=2))


def wrap_block(tag: str, payload: str) -> str:
    return f"<{tag}>\n{payload}\n</{tag}>"


_CODE_FENCE = re.compile(r"^```(?:json)?\s*([\s\S]*?)\s*```$", re.IGNORECASE)


def remove_code_fence(text: str) -> str:
    match = _CODE_FENCE.match(text)
    return match.group(1) if match else text


def read_context_files(relative_paths: Iterable[str], root: Path | None = None) -> str:
    """Read canonical instruction files into one delimited block.

    The path list is sorted so the block is byte-identical across stages and runs, which
    is required for DeepSeek prefix-cache hits.
    """
    base = root or ROOT
    parts = []
    for relative_path in sorted(set(relative_paths)):
        absolute = base / relative_path
        try:
            content = absolute.read_text(encoding="utf-8")
        except OSError as error:
            raise RunnerError(f"Required canonical context is missing: {relative_path}") from error
        parts.append(f'<document path="{relative_path}">\n{content}\n</document>')
    return "\n\n".join(parts)


_HEADING = re.compile(r"^(#{2})\s+(.*?)\s*$")


def extract_sections(document: str, heading_texts: Sequence[str]) -> tuple[dict[str, str], list[str]]:
    """Collect one or more named ``##`` sections from a canonical Markdown document.

    Extraction is by exact heading text, and a missing heading is reported to the caller
    instead of silently yielding an empty block, because a stage that silently receives no
    style contract writes prose against no standard at all.

    Returns ``(sections, missing)`` where ``sections`` maps the requested heading text to
    its extracted body.
    """
    lines = re.split(r"\r?\n", document)
    wanted = {text.strip().lower() for text in heading_texts}
    found: dict[str, str] = {}
    current: str | None = None
    buffer: list[str] = []

    def flush() -> None:
        nonlocal buffer
        if current is not None:
            found[current] = "\n".join(buffer).strip()
        buffer = []

    for line in lines:
        match = _HEADING.match(line)
        if match:
            title = f"## {match.group(2).strip()}"
            if title.lower() in wanted:
                flush()
                current = title
                buffer.append(line)
                continue
            if current is not None:
                # A different level-2 heading ends the wanted section.
                flush()
                current = None
                continue
        if current is not None:
            buffer.append(line)
    flush()
    missing = [heading for heading in heading_texts if heading.strip() not in found]
    return found, missing


def extract_context_sections(
    relative_path: str, heading_texts: Sequence[str], root: Path | None = None
) -> tuple[str, list[str]]:
    base = root or ROOT
    absolute = base / relative_path
    try:
        document = absolute.read_text(encoding="utf-8")
    except OSError as error:
        raise RunnerError(f"Required canonical context is missing: {relative_path}") from error
    sections, missing = extract_sections(document, heading_texts)
    ordered = [sections[heading.strip()] for heading in heading_texts if heading.strip() in sections]
    return "\n\n".join(ordered), missing


def next_attempt_directory(work_dir: Path) -> tuple[Path, int]:
    attempts_dir = work_dir / "attempts"
    attempts_dir.mkdir(parents=True, exist_ok=True)
    numbers = []
    for entry in attempts_dir.iterdir():
        if entry.is_dir():
            match = re.match(r"^attempt-(\d+)$", entry.name)
            if match:
                numbers.append(int(match.group(1)))
    attempt_number = (max(numbers) if numbers else 0) + 1
    attempt_dir = attempts_dir / f"attempt-{attempt_number}"
    attempt_dir.mkdir(parents=True, exist_ok=True)
    return attempt_dir, attempt_number


def validate_artifact_text(stage: Sequence[str], text: str, source_description: str) -> str:
    """Validate one stage's raw model output and return the artifact text.

    ``stage`` is ``(name, artifact, format, purpose)``, matching the JavaScript tuple.
    """
    name, _artifact, output_format = stage[0], stage[1], stage[2]
    artifact = text.strip()
    if not artifact:
        raise RunnerError(f"{source_description} is empty for stage {name}")
    if output_format == "JSON":
        artifact = remove_code_fence(artifact)
        try:
            json.loads(artifact)
        except ValueError as error:
            raise RunnerError(f"{source_description} is not valid JSON for stage {name}: {error}") from error
    return artifact


def relative_to_root(file_path: Path, root: Path | None = None) -> str:
    """A POSIX-style path relative to the repository root, as the run records use."""
    base = root or ROOT
    try:
        return file_path.resolve().relative_to(base.resolve()).as_posix()
    except ValueError:
        return file_path.as_posix()

File: src/runtime/artifacts.mjs
// Shared stage plumbing used by the editorial pipeline.
//
// Filesystem utilities, run-directory bookkeeping, Markdown section extraction, and
// artifact validation live here. DeepSeek transport, retry policy, and cost accounting
// have been moved to their own modules under `src/`; this module keeps the utilities
// every stage needs to read, write, and validate its artifacts.

import { cp, mkdir, readFile, readdir, stat, writeFile } from "node:fs/promises";
import path from "node:path";
import process from "node:process";
import { fileURLToPath } from "node:url";

export const ROOT = path.resolve(path.dirname(fileURLToPath(import.meta.url)), "..", "..");
export const RUNS_DIRECTORY = ".digest-runs";

// Reasoning tokens count against max_tokens on the DeepSeek model. A measured replay
// showed edit stages spending 28-31K reasoning tokens on a small corpus, which left only
// ~1.5K for the artifact and silently truncated it. The cap must therefore cover reasoning
// plus the full artifact. The model still generates only what it needs; this is a ceiling,
// not a target, so headroom is free. DeepSeek's maximum is 384K.
export const MAX_OUTPUT_TOKENS = Number(process.env.DIGEST_MAX_OUTPUT_TOKENS ?? 262_144);

export class RunnerError extends Error {}

export function option(name) {
  const index = process.argv.indexOf(name);
  return index === -1 ? undefined : process.argv[index + 1];
}

export function requiredOption(name) {
  const value = option(name);
  if (!value || value.startsWith("--")) throw new RunnerError(`Missing required option: ${name}`);
  return value;
}

export function validateRunId(runId) {
  if (path.basename(runId) !== runId || ["", ".", ".."].includes(runId)) {
    throw new RunnerError("Run ID must be a single directory name");
  }
}

// Whether a path exists, for a file *or* a directory.
//
// This used `readFile`, which fails on a directory. Two guards therefore never fired:
// `prepareReplay`'s refusal to replay into an existing run directory, and `verify-replay`'s
// missing-run check. The first is a data-loss path — a replay into an existing run id would
// have copied over that run's recorded corpus — so the helper now asks the filesystem a
// question that has an answer for both kinds of path.
export async function exists(filePath) {
  try {
    await stat(filePath);
    return true;
  } catch {
    return false;
  }
}

export async function readJson(filePath) {
  return JSON.parse(await readFile(filePath, "utf8"));
}

export async function copyFile(source, destination) {
  await mkdir(path.dirname(destination), { recursive: true });
  await cp(source, destination);
}

/**
 * Write an artifact, recreating its directory if it has disappeared.
 *
 * Used for failure records. A run directory can be deleted while a run is in flight — most
 * easily by a person who believes the run is finished — and the failure path must not then
 * replace the error being recorded with an `ENOENT` about the record itself.
 */
export async function writeArtifact(filePath, content) {
  await mkdir(path.dirname(filePath), { recursive: true });
  await writeFile(filePath, content, "utf8");
}

export function wrapBlock(tag, payload) {
  return `<${tag}>\n${payload}\n</${tag}>`;
}

export function removeCodeFence(text) {
  const match = text.match(/^```(?:json)?\s*([\s\S]*?)\s*```$/i);
  return match ? match[1] : text;
}

// Read canonical instruction files into one delimited block. The path list is sorted
// so the block is byte-identical across stages and runs, which is required for
// DeepSeek prefix-cache hits.
export async function readContextFiles(relativePaths) {
  const parts = [];
  for (const relativePath of [...new Set(relativePaths)].sort()) {
    const absolute = path.join(ROOT, relativePath);
    const content = await readFile(absolute, "utf8").catch(() => {
      throw new RunnerError(`Required canonical context is missing: ${relativePath}`);
    });
    parts.push(`<document path="${relativePath}">\n${content}\n</document>`);
  }
  return parts.join("\n\n");
}

// Collect one or more named `##` sections from a canonical Markdown document.
//
// Stage-specific context depends on being able to hand a stage *part* of a style
// file rather than the whole thing. Extraction is by exact heading text, and a
// missing heading is reported to the caller instead of silently yielding an empty
// block, because a stage that silently receives no style contract writes prose
// against no standard at all.
export function extractSections(document, headingTexts) {
  const lines = document.split(/\r?\n/);
  const wanted = new Set(headingTexts.map((text) => text.trim().toLowerCase()));
  const found = new Map();
  let current = null;
  let buffer = [];
  const flush = () => {
    if (current) found.set(current, buffer.join("\n").trim());
    buffer = [];
  };
  for (const line of lines) {
    const match = /^(#{2})\s+(.*?)\s*$/.exec(line);
    if (match) {
      const title = `## ${match[2].trim()}`;
      if (wanted.has(title.toLowerCase())) {
        flush();
        current = title;
        buffer.push(line);
        continue;
      }
      if (current) {
        // A different level-2 heading ends the wanted section.
        flush();
        current = null;
        continue;
      }
    }
    if (current) buffer.push(line);
  }
  flush();
  return { sections: found, missing: [...headingTexts].filter((h) => !found.has(h.trim())) };
}

export async function extractContextSections(relativePath, headingTexts) {
  const absolute = path.join(ROOT, relativePath);
  const document = await readFile(absolute, "utf8").catch(() => {
    throw new RunnerError(`Required canonical context is missing: ${relativePath}`);
  });
  const { sections, missing } = extractSections(document, headingTexts);
  const ordered = headingTexts
    .map((heading) => sections.get(heading.trim()))
    .filter(Boolean);
  return { text: ordered.join("\n\n"), missing, relativePath };
}

export async function nextAttemptDirectory(workDir) {
  const attemptsDir = path.join(workDir, "attempts");
  await mkdir(attemptsDir, { recursive: true });
  const entries = await readdir(attemptsDir, { withFileTypes: true });
  const numbers = entries
    .filter((entry) => entry.isDirectory())
    .map((entry) => /^attempt-(\d+)$/.exec(entry.name))
    .filter(Boolean)
    .map((match) => Number(match[1]));
  const attemptNumber = (numbers.length ? Math.max(...numbers) : 0) + 1;
  const attemptDir = path.join(attemptsDir, `attempt-${attemptNumber}`);
  await mkdir(attemptDir, { recursive: true });
  return { attemptDir, attemptNumber };
}

export async function validateArtifactText(stage, text, sourceDescription) {
  const [name, , outputFormat] = stage;
  let artifact = text.trim();
  if (!artifact) throw new RunnerError(`${sourceDescription} is empty for stage ${name}`);
  if (outputFormat === "JSON") {
    artifact = removeCodeFence(artifact);
    try {
      JSON.parse(artifact);
    } catch (error) {
      throw new RunnerError(`${sourceDescription} is not valid JSON for stage ${name}: ${error.message}`);
    }
  }
  return artifact;
}

// The canonical stage directory for a run, shared by the orchestrator and the
// measurement/replay readers so a stage's filesystem location has one definition.
export function stageDirectory(runId, stageName) {
  return path.join(ROOT, RUNS_DIRECTORY, runId, stageName);
}